### DISCUSSION TITLE
Make links intelligent?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5564] [Ideas]
Is the [LinkCatcherWidget](https://tiddlywiki.com/#LinkCatcherWidget), or some other UI widget, somehow involved when clicking/navigating via a regular [[Bracket link]]? Or is [[bracket link]] handling dealt with separately and there is no mechanism involved which normally has a UI?

**IFF** the LinkCatcherWidget, or some other widget, *is* involved when one clicks [[bracket links]], then 

I would like to create "intelligent links" by providing commands to the LinkCatcherWidget `actions` attribute to perform different things based on the [[string]] inside the brackets (e.g if it is a tiddler tagged Foo vs Bar).

I'm also wondering if it is possible to specify templates for how the `[[string]]` is presented, e.g like so `<li><<currentString>></li>` ....this, as you note, requires some reference to the current string. In the LinkCatcherWidget you can call for `navigateTo` but this is, if I understand, only for the `actions` attribute (...or?) [Edit: Specifically the LinkCatcherWidget cannot reasonably do this since it only comes into action when the link is clicked. Maybe there's some other widget involved in presenting bracket links.]

So, again, is the LinkCatcherWidget involved when clicking a regular [[bracketed link]]. Where can I modify it in that case?

It would be very useful to be able to conveniently type *regular* [[bracket links]] and have them behave *intelligently* i.e  depending *context*!

Any other ideas for how to achieve this - for [[bracket links]] - otherwise?

**Comment by saqimtiaz** [comment]
@twMat also have a look at https://tiddlywiki.com/prerelease/#MessageCatcherWidget which might be a better candidate to be extended to address this.

For both widgets I think the issue is that the text of the link is not available and from memory with the way the parser and widget tree works that may not be straight forward to address. This needs some investigation. I'm short on opportunity for it at the moment but hopefully this information is helpful. I'll investigate further when I get the chance.

**Reply by saqimtiaz** [reply]
On second thought extending the link widget to include the contents of the <a> element in the tm-navigate message might be a good approach, though this may not always be just text which should be kept in mind.

**Comment by twMat** [comment]
@saqimtiaz - I appreciate your answer but I wonder if either of the widgets you refer to are "baked into" the [[bracked link]] syntax? (There are too many system tiddlers named "<something>link" for me to tell which implements the bracket link syntax.) The main point I'm after is user convenience: There are situations when it just isn't sufficient to fiddle with widget syntax or even macro syntax such as when stressily taking live notes or when the mind is filled with poetry and typing - even thinking - `<$widget...` just kills the sparkle sprinkling angels...

**Reply by saqimtiaz** [reply]
I have no idea what this means:
"but I wonder if either of the widgets you refer to are "baked into" the [[bracked link]] syntax?"

You may be overthinking this. Irrespective of syntax, a link to a tiddler is a link to a tiddler, and all are implemented under the covers by the Link Widget. Try typing a link to a tiddler using different syntax and preview the parseTree in the editor in the prerelease.

The LinkCatcher and MessageCatcher widgets both allow you to execute actions in response to clicking a link, instead of the default behaviour. They have nothing to do with how that link was created but allow you to respond to clicks on links in content inside them.

Familiarizing yourself with the LinkCatcher and it's documentation may help in wrapping your head around it.

If the issue still isn't clear, hopefully someone else can chime in with a more in depth explanation.

**Comment by twMat** [comment]
>I have no idea what this means:
"but I wonder if either of the widgets you refer to are "baked into" the [[bracked link]] syntax?"

I mean that TW adds some layer to the primitive `<a href>` html syntax, to enable the bracket link syntax. And that at this TW layer there may be some hook or a way to customize how links behave. (I was thinking maybe there's some widget in that layer, comparable to how `<<macros>>` can have widgets inside.)

>a link to a tiddler, and all are implemented under the covers by the Link Widget. 

OK, so the [[bracket syntax]] is really a LinkWidget and, I assume, then defined in [widgets/link.js](https://tiddlywiki.com/prerelease/#%24%3A%2Fcore%2Fmodules%2Fwidgets%2Flink.js). So;

Would it be possible to hook into, or modify, the linkwidget so it also can perform actions so that the [[bracket syntax]] - without any other surrounding "catcherwidget" - could do intelligent stuff and behave differently depending on context? For example:

* ...Clicking on `[[Bob Ross]]` in the in the "Profiles" tiddler opens the "Bob Ross" tiddler but clicking on `[[Bob Ross]]` in the "Contact us" tiddler opens a "sendto" tiddler. 
* ...and clicking on `[[Bob Ross]]` when he is *tagged* with "Alert" does something special
* ...and clicking on `[[bobross]]` also opens the "Bob Ross" tiddler because I've made missing links cleverly guess tiddler or look them up in a dictionary.

Again, this is all doable enough with *surrounding* widgets, but I hope to modify the behaviour of the plain `[[bracket link]]` itself because it would be extremely smooth to just predefine such "behaviours" and then only type the plain bracketed link, instead of having to mess with widgets or macros to invoke the contextual behaviours.

**Comment by pmario** [comment]
>Is the LinkCatcherWidget, or some other UI widget, somehow involved when clicking/navigating via a regular [[Bracket link]]? Or is [[bracket link]] handling dealt with separately and there is no mechanism involved which normally has a UI?

No. ... The regular link emits a `tm-navigate` message, that is "catched" by the `<$navigator` widget. ... The link-catcher can "catch" this message and manipulate or even "block" it and do something completely different.

**Comment by pmario** [comment]
Hi Mat, 
My uni-link plugin uses a macro instead of creating a link-widget. see: https://wikilabs.github.io/editions/uni-link/#%24%3A%2Fplugins%2Fwikilabs%2Funi-link%2Funi-link-macro  ... 

So if you write [[this is a link]] it actually calls `<<uni-link "this is a link">>`. The advantage is, that there is some code, that will track the backlinks. 

It may be possible to modify the uni-link-macro code and see, what happens. You can play with it.

**Reply by pmario** [reply]
Also see this issue at: https://github.com/wikilabs/plugins/issues/40 ... Where I'm thinking about a new possibility to create transclusions if you write eg: `[[link]>template] .. almost same as .. {{link||template}}`

**Comment by twMat** [comment]
@pmario - WHOAAAA!!!!!

>So if you write [[this is a link]] it actually calls <<uni-link "this is a link">>.

You have made it in wikitext (not js) and even my very first attempt - (while thinking "Of course this isn't gonna work") - succeeded!!! It seems to be possible to make exactly what I ask for with it. Big thank you!

This will totally ruin my workday tomorrow ;-) ;-) ;-)

**Reply by pmario** [reply]
It contains some JS code. It actually replaces the "prettylink" parsing rules. ... But I wanted to be able to style links differently. eg: title, caption and subtitle can have a different colour. .. That was one of the reasons, to use macro calls 

... Be careful -- you can have an impact on site performance, if you use too many "heavy weight" filters in the macros.

**Comment by linonetwo** [comment]
This seems simillar to https://github.com/Jermolene/TiddlyWiki5/issues/4982 ? Smart link - Typed Link
